DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/ip_authoring_mmhost/src/mmhost.cpp

```diff
@@ -0,0 +1,184 @@
+      int mul = x[i] * y[i];
+      int add = mul + z[i];
+      y[i] = mul;
+      z[i] = add;
+    }
+  }
+};
+
+// Using mmhost macro
+struct VectorMADIP {
+  mmhost(BL1,  // buffer_location or aspace
+         28,   // address width
+         64,   // data width
+         16,   // ! latency, must be atleast 16
+         1,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+         1,    // maxburst
+         0,    // align, 0 defaults to alignment of the type
+         1     // waitrequest, 0: false, 1: true
+         ) int *x;
+  mmhost(BL2,  // buffer_location or aspace
+         28,   // address width
+         64,   // data width
+         16,   // ! latency, must be atleast 16
+         0,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+         1,    // maxburst
+         0,    // align, 0 defaults to alignment of the type
+         1     // waitrequest, 0: false, 1: true
+         ) int *y;
+  mmhost(BL3,  // buffer_location or aspace
+         28,   // address width
+         64,   // data width
+         16,   // ! latency, must be atleast 16
+         0,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+         1,    // maxburst
+         0,    // align, 0 defaults to alignment of the type
+         1     // waitrequest, 0: false, 1: true
+         ) int *z;
+  int size;
+
+  VectorMADIP(int *x_, int *y_, int *z_, int size_)
+      : x(x_), y(y_), z(z_), size(size_) {}
+
+  void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      int mul = x[i] * y[i];
+      int add = mul + z[i];
+      y[i] = mul;
+      z[i] = add;
+    }
+  }
+};
+
+void fillArrays(int *x, int *y, int *z, int size) {
+  for (int i = 0; i < size; ++i) {
+    x[i] = i;
+    y[i] = i * 2;
+    z[i] = i * 3;
+  }
+}
+
+int main(void) {
+#if FPGA_SIMULATOR
+  auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+  auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+  auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+  try {
+    sycl::queue q(selector, fpga_tools::exception_handler,
+                  property::queue::enable_profiling{});
+
+    // Print out the device information.
+    std::cout << "Running on device: "
+              << q.get_device().get_info<info::device::name>() << "\n";
+
+    int size = 10000;
+    double start, end;
+    event e;
+
+    // MMHost arguments
+    auto x = make_malloc_shared<int>(q, size, BL1);
+    auto y = make_malloc_shared<int>(q, size, BL2);
+    auto z = make_malloc_shared<int>(q, size, BL3);
+
+    fillArrays(x.get(), y.get(), z.get(), size);
+
+    e = q.single_task(VectorMADIP{x.get(), y.get(), z.get(), size});
+    e.wait();
+
+    start = e.get_profiling_info<info::event_profiling::command_start>();
+    end = e.get_profiling_info<info::event_profiling::command_end>();
+    // convert from nanoseconds to ms
+    double kernel_mmhost_time = (double)(end - start) * 1e-6;
+
+    // Regular pointer arguments
```

Comment:
```suggestion
```suggestion
    // Allocate memory for pointer kernel arguments, no buffer location is specified
```
```